arduino/sketches: build sketches as a module

cladmi opened the pull request:
Generate a module for arduino sketches in a subfolder of BINDIR.
This prevents issues when doing concurrent builds or out of tree build with readonly sources.

This fixes #5848

The concept is instead of generating a cpp file in the source directory, it is generated into the build directory. However, this requires defining a new module for this directory.
It is now generated during build and not parsing.
I did not addressed the "write atomically" as I think it is not necessary anymore.

To do this, I replaced the previous bash script by a Makefile script which creates the module as `link` target prerequisite.


### Testing procedure

As it was not asked before, I will give a testing procedure:

Setup:

* Clean
*    create bin
*    remove write permission in the application directory

```
git clean -xdf examples/arduino_hello-world
mkdir -p examples/arduino_hello-world/bin
chmod -w examples/arduino_hello-world
make -C examples/arduino_hello-world clean all BOARD=arduino-mega2560
```

Cleanup

* `chmod u+w examples/arduino_hello-world`

It works with this PR.

It fails on master, as we try to write in the directory and so the main function is never declared.
You can also notice that the Permission denied are there before even starting to build. make clean also tries to create the file.

Fixes #5848

smlng commented on sys/arduino/doc.txt:
```diff
@@ -64,6 +64,8 @@
  * [Makefile.include](https://github.com/RIOT-OS/RIOT/tree/master/sys/arduino/Makefile.include)
  * of the RIOT Arduino module.
  *
+ * TODO UPDATE DOCUMENTATION
```
would be very much appreciated 😄

cladmi replied:
Yeah, thats why I put this line either I could forget.

cladmi commented:
Finally updated the documentation.
I rebased with the current `master` branch.

My questions are:
* what do you think about the concept?
* do you think it would be better that `Makefile.arduino_sketches.inc` defined a function that we call after to create the rules or defining stuff directly in it is fine?

gebart commented on sys/arduino/Makefile.arduino_sketches.inc:
```diff
@@ -0,0 +1,31 @@
+	@mkdir -p $@
+
+# Make everything rebuild if current makefile changes
+CURMK = $(word $(words $(MAKEFILE_LIST)),$(MAKEFILE_LIST))
```
You could use the lastword function instead

gebart replied:
Better use a different name than CURMK. This file is included from the Makefile.include which means it will pollute the name space.

cladmi replied:
I will change to `lastword` I learned it after writing this, and update to a namespaced variable name.

cladmi commented:
I will change the Makefile name too. Do I just add `.mk` or do you have a better naming idea?

However, I do not want to move it to `makefiles`. I think the content of `makefiles` should only contain files that are related to the general build system, but not module specific stuff.

gebart commented on sys/arduino/arduino_sketches.inc.mk:
```diff
@@ -0,0 +1,31 @@
+	@mkdir -p $@
+
+# Make everything rebuild if current makefile changes
+_ARDUINO_SKETCHES_MAKEFILE = $(lastword $(MAKEFILE_LIST))
```
I think this needs an immediate assignment (`:=`)to avoid the situation where MAKEFILE_LIST is updated by including another file later.

cladmi replied:
The value is used directly and only in the two following lines, so following imports will not impact.
However, as I am not using immediate assignment, `lastword` is executed twice, so I will change.

gebart commented:
@cladmi sounds reasonable. :+1: for arduino_sketches.inc.mk

cladmi commented:
Can I rebase and squash ?

gebart commented on sys/arduino/arduino_sketches.inc.mk:
```diff
@@ -0,0 +1,31 @@
+# Compile together the Arduino sketches of the application
+# They are declared a as new module $(SKETCH_MODULE) in $(SKETCH_MOD_DIR)
```
typo SKETCH_MOD_DIR -> SKETCH_MODULE_DIR

cladmi replied:
Thanks, I missed this one.

gebart commented on sys/arduino/arduino_sketches.inc.mk:
```diff
@@ -0,0 +1,31 @@
+# Building the module files
+#   Do not use $^ in receipes as Makefile is also a prerequisite
+$(SKETCH_MODULE_DIR)/Makefile: $(SKETCH_MODULE_DIR)/$(SKETCH_CPP)
+	$(Q)echo 'SRCXX = $(SKETCH_CPP)'               > $@
```
Is there any way this recipe can be called multiple times during the same build? (race condition clobbering the destination file)
For example if building multiple sketches in different source directories?

cladmi replied:
You cannot put sketches in multiple directories as `post.snip` is defining the `main` function.
As I understood it, its to replace the code you would put in your application directory by this arduino sketches thing. That's why SKETCHES are in `APPDIR` which is basically `CURDIR`.

gebart commented:
OK to squash

cladmi commented:
@smlng @haukepetersen do you agree with this version ?

smlng commented:
untested ACK

gebart commented:
I forgot about this one, I think it is pretty much ready for merge.

cladmi commented:
I noticed this one suffers from the same problem as all modules where deleting one of the `SKETCHES` would not trigger a rebuild.
I will force re-building the `.cpp` everytime to get the same behaviour as before.

There are different ways to properly fix it:
 * either generate an equivalent of `.d` file included by make
 * change the `sketches.cpp` to directly include the `.sketch` files and GCC would do the job of generating the `.d` file.

cladmi commented:
I needs to modify this as now "link" is not the right target to depend anymore

cladmi commented:
I hacked it the same way `javascript` and `lua` are doing by making it build before RIOTBUILD_CONFIG_HEADER_C.

It could be fixed after https://github.com/RIOT-OS/RIOT/pull/9002

cladmi commented:
Changed to use `BUILDDEPS`.

jia200x commented:
@cladmi any news with this?

cladmi commented:
I changed some things since @gebart ACKed the PR as the missing `BUILDDEPS` was merged.
So it needs to be re-checked and ACKed and say when I can squash.

 I rebased to master. 


### Testing procedure

As it was not asked before, I will give a testing procedure:

Setup:
* Clean
* create `bin` 
* remove write permission in the application directory

```
git clean -xdf examples/arduino_hello-world
mkdir -p examples/arduino_hello-world/bin
chmod -w examples/arduino_hello-world
```
Then try to compile for an arduino board
```
make -C examples/arduino_hello-world clean all BOARD=arduino-mega2560
```

Cleanup
```
chmod u+w examples/arduino_hello-world
```

It works with this PR.

It fails on master, as we try to write in the directory and so the `main` function is never declared.
You can also notice that the `Permission denied` are there before even starting to build. `make clean` also tries to create the file.

miri64 commented:
This was ACK'd almost a year ago. What's the hold-up?

cladmi commented:
As my previous comment said, as other PRs were merged I changed things in this one after it was ACK.

And I just noticed that, as we had the issue recently in another PR, the directory target will not work on mac as is as it is using `make 3.81`.
It may also have issues with parallel `make -j clean all` execution.

I will look into these ones when I find time.

cladmi commented:
I removed the directory creation target and included it in the `.cpp` file creation.
I also now declare all generated files as `BUILDDEPS` to prevent any issues with `clean all` in parallel.

The handling of re-building is still done with a `FORCE` and doing it with `.d` files would be better handled globally.
I would squash the two commits together at the end.

tcschmidt commented:
@smlng this update needs a review

smlng commented:
also: please squash

cladmi commented:
> One question: why `sys/arduino/arduino_sketches.inc.mk` and not `makefiles/arduino/arduino_sketches.inc.mk` - and (as it is already in a subdir) call the file `sketches.inc.mk`?

I find it really stupid as organization to not be in the `sys/arduino` directory.
Having a module specific handling in `makefiles` is putting specific code in your general purpose library.

For the `arduino_sketches`, I named it as I created the module with the name `arduino_sketches`. But indeed, there is already the `arduino` directory. I will push the change first (so you can see the diff) and then squash.

cladmi commented:
Why is it not showing the commit in the text description? it was this one to help you review: https://github.com/RIOT-OS/RIOT/commit/c3035dfb12e41599c6b65a7feef3769dd1d072eb

smlng commented:
> I find it really stupid as organization to not be in the `sys/arduino` directory.
> Having a module specific handling in `makefiles` is putting specific code in your general purpose library.


agreed, thanks for the explanation!

cladmi commented:
Thank you for the review.

This now does not generate files while parsing dependencies anymore !
